govcd/cse.go

```diff
@@ -0,0 +1,379 @@
+	// "fighting" with the CSE Server
+	retries := 0
+	maxRetries := 5
+	updated := false
+	for retries <= maxRetries {
+		rde, err := getRdeById(cluster.client, cluster.ID)
+		if err != nil {
+			return err
+		}
+
+		rde.DefinedEntity.Entity = entityContent
+		err = rde.Update(*rde.DefinedEntity)
+		if err == nil {
+			updated = true
+			break
+		}
+		if err != nil {
+			// If it's an ETag error, we just retry without waiting
+			if !strings.Contains(strings.ToLower(err.Error()), "etag") {
+				return err
+			}
+		}
+		retries++
+		util.Logger.Printf("[DEBUG] The request to update the Kubernetes cluster '%s' failed due to a ETag lock. Trying again", cluster.ID)
+	}
+
+	if !updated {
+		return fmt.Errorf("could not update the Kubernetes cluster '%s' after %d retries, due to an ETag lock blocking the operations", cluster.ID, maxRetries)
+	}
+
+	return cluster.Refresh()
+}
+
+// Delete deletes a CSE Kubernetes cluster, waiting the specified amount of minutes. If the timeout is reached, this method
+// returns an error, even if the cluster is already marked for deletion.
+func (cluster *CseKubernetesCluster) Delete(timeoutMinutes time.Duration) error {
+	var elapsed time.Duration
+	start := time.Now()
+	markForDelete := false
+	forceDelete := false
+	for elapsed <= timeoutMinutes*time.Minute || timeoutMinutes == 0 { // If the user specifies timeoutMinutes=0, we wait forever
+		rde, err := getRdeById(cluster.client, cluster.ID)
+		if err != nil {
+			if ContainsNotFound(err) {
+				return nil // The RDE is gone, so the process is completed and there's nothing more to do
+			}
+			return fmt.Errorf("could not retrieve the Kubernetes cluster with ID '%s': %s", cluster.ID, err)
+		}
+
+		markForDelete = traverseMapAndGet[bool](rde.DefinedEntity.Entity, "spec.vcdKe.markForDelete")
+		forceDelete = traverseMapAndGet[bool](rde.DefinedEntity.Entity, "spec.vcdKe.forceDelete")
+
+		if !markForDelete || !forceDelete {
+			// Mark the cluster for deletion
+			rde.DefinedEntity.Entity["spec"].(map[string]interface{})["vcdKe"].(map[string]interface{})["markForDelete"] = true
+			rde.DefinedEntity.Entity["spec"].(map[string]interface{})["vcdKe"].(map[string]interface{})["forceDelete"] = true
+			err = rde.Update(*rde.DefinedEntity)
+			if err != nil {
+				// We ignore any ETag error. This just means a clash with the CSE Server, we just try again
+				if !strings.Contains(strings.ToLower(err.Error()), "etag") {
+					return fmt.Errorf("could not mark the Kubernetes cluster with ID '%s' to be deleted: %s", cluster.ID, err)
+				}
+			}
+		}
+
+		util.Logger.Printf("[DEBUG] Cluster '%s' is still not deleted, will check again in 10 seconds", cluster.ID)
+		time.Sleep(10 * time.Second)
+		elapsed = time.Since(start)
+	}
+
+	// We give a hint to the user about the deletion process result
+	if markForDelete && forceDelete {
+		return fmt.Errorf("timeout of %v minutes reached, the cluster was successfully marked for deletion but was not removed in time", timeoutMinutes)
```

Comment:
```suggestion
		return fmt.Errorf("timeout of %s reached, the cluster was successfully marked for deletion but was not removed in time", timeout)
```
instances of `time.Deletion` convert to string in a readable way

Reply:
Fixed